nilearn/decoding/space_net_solvers.py

```diff
@@ -0,0 +1,539 @@
+        w = w[:-1]
+
+    grad_id = _gradient_id(_unmask(w, mask), l1_ratio=l1_ratio)
+    out += alpha * _tvl1_objective_from_gradient(grad_id)
+
+    return out
+
+
+def tvl1_solver(X, y, alpha, l1_ratio, mask, loss=None, max_iter=100,
+                lipschitz_constant=None, init=None,
+                prox_max_iter=5000, tol=1e-4, callback=None, verbose=1):
+    """Minimizes empirical risk for TV-L1 penalized models.
+
+    Can handle least squares (mean squared error --a.k.a mse) or logistic
+    regression. The same solver works for both of these losses.
+
+    This function invokes the mfista backend (from fista.py) to solver the
+    underlying optimization problem.
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_samples, n_features)
+        Design matrix.
+
+    y : ndarray, shape (n_samples,)
+        Target / response vector.
+
+    alpha : float
+        Constant that scales the overall regularization term. Defaults to 1.0.
+
+    l1_ratio : float in the interval [0, 1]; optinal (default .5)
+        Constant that mixes L1 and TV penalization.
+        l1_ratio == 0 : just smooth. l1_ratio == 1 : just lasso.
+        Defaults to 0.5.
+
+    mask : ndarray, shape (nx, ny, nz)
+        The support of this mask defines the ROIs being considered in
+        the problem.
+
+    max_iter : int
+        Defines the iterations for the solver. Defaults to 100
+
+    prox_max_iter : int, optional (default 5000)
+        Maximum number of iterations for inner FISTA loop in which
+        the prox of TV is approximated.
+
+    tol : float
+        Defines the tolerance for convergence. Defaults to 1e-4.
+
+    loss : string
+        Loss model for regression. Can be "mse" (for squared loss) or
+        "logistic" (for logistic loss).
+
+    lipschitz_constant : float, optional (default None)
+        Lipschitz constant (i.e an upper bound of) of gradient of smooth part
+        of the energy being minimized. If no value is specified (None),
+        then it will be calculated.
+
+    callback : callable(dict) -> bool, optional (default None)
+        Function called at the end of every energy descendent iteration of the
+        solver. If it returns True, the loop breaks.
+
+    Returns
+    -------
+    w : ndarray, shape (n_features,)
+       The solution vector (Where `w_size` is the size of the support of the
+       mask.)
+
+    objective : array of floats
+        Objective function (fval) computed on every iteration.
+
+    solver_info: float
+        Solver information, for warm start.
+
+    """
+
+    # sanitize loss
+    if loss not in ["mse", "logistic"]:
+        raise ValueError("'%s' loss not implemented. Should be 'mse' or "
+                         "'logistic" % loss)
+
+    # shape of image box
+    flat_mask = mask.ravel()
+    volume_shape = mask.shape
+
+    # We'll work on the full brain, and do the masking / unmasking
+    # magic when needed
+    w_size = X.shape[1] + int(loss == "logistic")
```

Comment:
comment about intercept and preferably make this more explicit for readability by future contributors.